Simplify process(Script|Python)(Remote)?Call

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58753 Fix race condition in TP agent
* #58675 Make Future always extract DataPtrs in RPC tests
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef
* #58428 Make TP agent use streams from Future when sending response
* #58674 Set and propagate devices in RRef completion future
* #58427 Set streams when invoking UDFs
* #58426 Create CUDA-aware futures in RequestCallback
* #58425 Provide pre-extracted DataPtrs when completing a Future with a Message
* #58424 Allow Future::then to return pre-extracted DataPtrs
* #58423 Always use intrusive_ptr for Message (2 out of 2)
* #58422 Always use intrusive_ptr for Message (1 out of 2)
* #58421 Prevent using anything other than intrusive_ptr for Future
* #58420 Migrate remaining shared_ptr<Future> to intrusive_ptr
* #57861 Make remaining autograd methods return futures
* #57860 Make remaining RRef methods return futures
* #57859 Unify fetching RRefs
* #57858 Deduplicate Python object serialization
* **#57857 Simplify process(Script|Python)(Remote)?Call**
* #57856 Unify assignment of OwnerRRef result
* #57855 Make processPythonExecution return a future
* #57854 Remove getScriptRemoteCallType
* #57853 Simplify OwnerRRef completion
* #57852 Unify async execution for JIT functions
* #57851 Unify invoking JIT functions
* #57850 Unify invoking JIT operands
* #57849 Make some methods of RequestCallback return void instead of bool
* #57848 Centralize setting messageId in RequestCallback
* #57847 Make RequestCallback collect Futures from methods, rather than providing them
* #57846 Add helpers to manipulate futures
* #57638 Reduce overhead when Future invokes callbacks inline
* #57637 Introduce thenAsync method on Future

There used to be a whole lot of methods: `processPythonCall`, `processScriptCall`, `processScriptRemoteCall`, `processPythonRemoteCall`, `processScriptCallOp`, `processBaseScriptRemoteCall` and `processScriptRemoteCallOp`. Thanks to the previous simplification, we can now drop all but the first four, which map nicely 1:1 to the four message types we need to handle. Also their signatures become much simpler: they take an RPC command and return a future.

Differential Revision: [D28253848](https://our.internmc.facebook.com/intern/diff/D28253848/)

lw committed:
Simplify process(Script|Python)(Remote)?Call

There used to be a whole lot of methods: `processPythonCall`, `processScriptCall`, `processScriptRemoteCall`, `processPythonRemoteCall`, `processScriptCallOp`, `processBaseScriptRemoteCall` and `processScriptRemoteCallOp`. Thanks to the previous simplification, we can now drop all but the first four, which map nicely 1:1 to the four message types we need to handle. Also their signatures become much simpler: they take an RPC command and return a future.

Differential Revision: [D28253848](https://our.internmc.facebook.com/intern/diff/D28253848/)

[ghstack-poisoned]

lw committed:
Update on "Simplify process(Script|Python)(Remote)?Call"


There used to be a whole lot of methods: `processPythonCall`, `processScriptCall`, `processScriptRemoteCall`, `processPythonRemoteCall`, `processScriptCallOp`, `processBaseScriptRemoteCall` and `processScriptRemoteCallOp`. Thanks to the previous simplification, we can now drop all but the first four, which map nicely 1:1 to the four message types we need to handle. Also their signatures become much simpler: they take an RPC command and return a future.

Differential Revision: [D28253848](https://our.internmc.facebook.com/intern/diff/D28253848/)

[ghstack-poisoned]

lw committed:
Update on "Simplify process(Script|Python)(Remote)?Call"


There used to be a whole lot of methods: `processPythonCall`, `processScriptCall`, `processScriptRemoteCall`, `processPythonRemoteCall`, `processScriptCallOp`, `processBaseScriptRemoteCall` and `processScriptRemoteCallOp`. Thanks to the previous simplification, we can now drop all but the first four, which map nicely 1:1 to the four message types we need to handle. Also their signatures become much simpler: they take an RPC command and return a future.

Differential Revision: [D28253848](https://our.internmc.facebook.com/intern/diff/D28253848/)

[ghstack-poisoned]

lw committed:
Update on "Simplify process(Script|Python)(Remote)?Call"


There used to be a whole lot of methods: `processPythonCall`, `processScriptCall`, `processScriptRemoteCall`, `processPythonRemoteCall`, `processScriptCallOp`, `processBaseScriptRemoteCall` and `processScriptRemoteCallOp`. Thanks to the previous simplification, we can now drop all but the first four, which map nicely 1:1 to the four message types we need to handle. Also their signatures become much simpler: they take an RPC command and return a future.

Differential Revision: [D28253848](https://our.internmc.facebook.com/intern/diff/D28253848/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -137,74 +137,75 @@ std::unique_ptr<RpcCommandBase> RequestCallbackImpl::
+c10::intrusive_ptr<JitFuture> RequestCallbackImpl::processPythonCall(
+    RpcCommandBase& rpc) const {
   auto& upc = static_cast<UnpickledPythonCall&>(rpc);
   auto future = runPythonFunction(upc.pythonUdf(), upc.isAsyncExecution());
```
I assume, later, we will need to pass agent devices to this Future?

lw replied:
Correct, I'm doing it in #58426.

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -137,74 +137,75 @@ std::unique_ptr<RpcCommandBase> RequestCallbackImpl::
-    responseFuture->markCompleted(c10::make_intrusive<Message>(
-        PythonResp(std::move(serializedPyObj)).toMessage()));
-  });
+  return future->then(
```
Each of the futures in the chain will introduce another data ptrs extraction for CUDA (when CUDA is enabled here). Do we need to worry about the perf issue here?

lw replied:
I have been asking myself the same thing. I'm somewhat worried about DataPtr extraction (although not too much, because for the most part here we're inspecting JIT IValues, which should be much cheaper than inspecting Python objects) but I'm also a bit worried about creating/recording/waiting on events each time, getting new streams, ...

I have some ideas on how we can mitigate some of these issues. For example in #58424 I'll allow callbacks to provide pre-extracted DataPtrs, which should allow us to skip extraction in some cases. I also believe we can add some fastpaths in Future so that for example we reuse the same streams from the "parent" when running callbacks.

Honestly, though, I'm also fine for now to take a perf hit if necessary, if that means we can get CUDA support to work correctly and reliably, and then look later into optimizing it.

mrshenli reviewed (approved):
LGTM

lw committed:
Update on "Simplify process(Script|Python)(Remote)?Call"


There used to be a whole lot of methods: `processPythonCall`, `processScriptCall`, `processScriptRemoteCall`, `processPythonRemoteCall`, `processScriptCallOp`, `processBaseScriptRemoteCall` and `processScriptRemoteCallOp`. Thanks to the previous simplification, we can now drop all but the first four, which map nicely 1:1 to the four message types we need to handle. Also their signatures become much simpler: they take an RPC command and return a future.

Differential Revision: [D28253848](https://our.internmc.facebook.com/intern/diff/D28253848/)

[ghstack-poisoned]

lw committed:
Update on "Simplify process(Script|Python)(Remote)?Call"


There used to be a whole lot of methods: `processPythonCall`, `processScriptCall`, `processScriptRemoteCall`, `processPythonRemoteCall`, `processScriptCallOp`, `processBaseScriptRemoteCall` and `processScriptRemoteCallOp`. Thanks to the previous simplification, we can now drop all but the first four, which map nicely 1:1 to the four message types we need to handle. Also their signatures become much simpler: they take an RPC command and return a future.

Differential Revision: [D28253848](https://our.internmc.facebook.com/intern/diff/D28253848/)

[ghstack-poisoned]

lw committed:
Update on "Simplify process(Script|Python)(Remote)?Call"


There used to be a whole lot of methods: `processPythonCall`, `processScriptCall`, `processScriptRemoteCall`, `processPythonRemoteCall`, `processScriptCallOp`, `processBaseScriptRemoteCall` and `processScriptRemoteCallOp`. Thanks to the previous simplification, we can now drop all but the first four, which map nicely 1:1 to the four message types we need to handle. Also their signatures become much simpler: they take an RPC command and return a future.

Differential Revision: [D28253848](https://our.internmc.facebook.com/intern/diff/D28253848/)

[ghstack-poisoned]

lw committed:
Update on "Simplify process(Script|Python)(Remote)?Call"


There used to be a whole lot of methods: `processPythonCall`, `processScriptCall`, `processScriptRemoteCall`, `processPythonRemoteCall`, `processScriptCallOp`, `processBaseScriptRemoteCall` and `processScriptRemoteCallOp`. Thanks to the previous simplification, we can now drop all but the first four, which map nicely 1:1 to the four message types we need to handle. Also their signatures become much simpler: they take an RPC command and return a future.

Differential Revision: [D28253848](https://our.internmc.facebook.com/intern/diff/D28253848/)

[ghstack-poisoned]
